drivers/i3c/i3c_npcx.c

```diff
@@ -958,35 +1043,63 @@ static int npcx_i3c_transfer(const struct device *dev, struct i3c_device_desc *t
 		}
 #endif
 
-		/*
-		 * Two ways to do read/write transfer .
-		 * 1. [S] + [0x7E]    + [address] + [data] + [Sr or P]
-		 * 2. [S] + [address] + [data]    + [Sr or P]
-		 *
-		 * Send broadcast header(0x7E) on first transfer or after a STOP,
-		 * unless flag is set not to.
-		 */
-		if (!(msgs[i].flags & I3C_MSG_NBCH) && (send_broadcast)) {
-			ret = npcx_i3c_request_emit_start(inst, I3C_BROADCAST_ADDR,
-							  NPCX_I3C_MCTRL_TYPE_I3C, false, 0);
-			if (ret < 0) {
-				LOG_ERR("%s: emit start of broadcast addr failed, error (%d)",
-					__func__, ret);
+		/* Check message SDR or HDR mode */
+		bool is_msg_hdr = (msgs[i].flags & I3C_MSG_HDR) == I3C_MSG_HDR;
+
+		/* Set emit start type SDR or HDR-DDR mode */
+		if (!is_msg_hdr || msgs[i].hdr_mode == 0) {
+			op_type = NPCX_I3C_MCTRL_TYPE_I3C; /* Set operation type SDR */
+
+			/*
+			 * SDR, send boradcast header(0x7E)
+			 *
+			 * Two ways to do read/write transfer (SDR mode).
+			 * 1. [S] + [0x7E]    + [address] + [data] + [Sr or P]
+			 * 2. [S] + [address] + [data]    + [Sr or P]
+			 *
+			 * Send broadcast header(0x7E) on first transfer or after a STOP,
+			 * unless flag is set not to.
+			 */
+			if (!(msgs[i].flags & I3C_MSG_NBCH) && send_broadcast) {
+				ret = npcx_i3c_request_emit_start(inst, I3C_BROADCAST_ADDR,
+								  NPCX_I3C_MCTRL_TYPE_I3C, false,
+								  0);
+				if (ret < 0) {
+					LOG_ERR("%s: emit start of broadcast addr failed, error "
+						"(%d)",
+						__func__, ret);
+					break;
+				}
+				send_broadcast = false;
+			}
+		} else if ((data->common.ctrl_config.supported_hdr & I3C_MSG_HDR_DDR) &&
+			   (msgs[i].hdr_mode == I3C_MSG_HDR_DDR) && is_msg_hdr) {
+
+			op_type = NPCX_I3C_MCTRL_TYPE_I3C_HDR_DDR; /* Set operation type DDR */
+
+			/* Check HDR-DDR moves data by words */
+			if ((msgs[i].len % 2) != 0x0) {
+				LOG_ERR("HDR-DDR data length should be number of words , xfer "
+					"len=%d", msgs[i].num_xfer);
+				ret = -EINVAL;
 				break;
 			}
-			send_broadcast = false;
+		} else {
+			LOG_ERR("%s: Controller HDR Mode %d", dev->name,
+				data->common.ctrl_config.supported_hdr);
+			LOG_ERR("%s: msg HDR Mode %#x, msg flag %#x", __func__, msgs[i].hdr_mode,
+				msgs[i].flags);
```

Comment:
2 error logs here??, also one looks like it's using dev->name before the `:` while the other is using `__func__`

Reply:
Done